### PR TITLE
[WIP] Add dear imgui

### PR DESCRIPTION
**Pull request by bbirchnz**
Add Dear IMGUI layer.

- [x]  Add Dear IMGUI to build
- [ ] Create class for layer that can render IMGUI, managing own context
- [ ] Add events handlers/custom IMGUI backend for mouse/tablet events
- [ ] Add lua bindings, lua context, and create a IMGUI/Lua tab type where custom interactive tabs can be built by user supplied Lua.

### CODE REVIEW

File: src/app/app-common/include/OpenKneeboard/ImguiInstance.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  void Render(ID2D1DeviceContext*, PageID, const D2D1_RECT_F& targetRect);
+
+ private:
+  DXResources mDXR;
+  KneeboardState* mKneeboard;
+  ImGuiContext* mImguiCtx;
+
+  winrt::com_ptr<ID2D1DeviceContext> m2DContext;
+  winrt::com_ptr<ID3D11Texture2D> mTexture;
+  winrt::com_ptr<ID2D1Bitmap1> mBitmap;
```
Comment (fredemmott) [inline src/app/app-common/include/OpenKneeboard/ImguiInstance.h:29]: Is this just for testing? I'm not certain it's possible, but I'm hoping the final version can operate directly on the same D2D contexts/textures that are used for the other rendering passes.

Also if this is going to stay around, make sure the pointer members above are initialized to nullptr

**Review by fredemmott — COMMENTED** [review]
Thanks, good start - left a few nits :)

File: src/app/app-common/include/OpenKneeboard/ImguiInstance.h (hunk 1 above)
Comment (bbirchnz) [reply src/app/app-common/include/OpenKneeboard/ImguiInstance.h:29]: Doesn't seem so, looks like Dear ImGui expects to render to a 3D context, this was the only way I could find to get it on a 2D Texture and control where it was rendered.

Comment (fredemmott) [reply src/app/app-common/include/OpenKneeboard/ImguiInstance.h:29]: Some options:
- The current surfaces are always backed by D3D; perhaps GetTarget().as<ID2D1Bitmap>().GetSurface(), though I’m not sure if that queryinterface is valid
- When setting the D2D context’s target, we could also set the underlying D3D implicit context’s target
- … worst case, could thread through the bitmap/DXGI surface/d3d texture

Comment (fredemmott) [reply src/app/app-common/include/OpenKneeboard/ImguiInstance.h:29]: There’s two sources for the usual textures:
- InterprocessRenderer (fully under our control)
- TabPage: partially under our control, but winui3 swapchainpanel is also involved here 

It’s also probably worth noting that these can have different sizes and viewports for the render

File: src/app/app-common/ImguiInstance.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+  mDXR.mD2DDeviceContext->CreateBitmapFromDxgiSurface(
+    mTexture.as<IDXGISurface>().get(), nullptr, mBitmap.put());
+
+  // we're gonna change render targets and generally screw with the state to
+  // render out IMGUI, these helpers should help us restore the state afterwards
+  vrperfkit::D3D11State state {};
+  vrperfkit::StoreD3D11State(ctx.get(), state);
+  scope_guard restoreState(
+    [&]() { vrperfkit::RestoreD3D11State(ctx.get(), state); });
```
Comment (fredemmott) [inline src/app/app-common/ImguiInstance.cpp:45]: I've removed vrperfkit from master, as OpenXR Toolkit has a more complete version of this functionality.

I've wrapped this up as `OpenKneeboard::D3D11::SavedState` - just create an instance, and it'll automatically restore when destructed.

Comment (fredemmott) [reply src/app/app-common/ImguiInstance.cpp:45]: changed again to use a D3D11 feature I didn't know about (thanks mbucchia :) ) , but no change to the internal API

**Comment by fredemmott** [comment]
Closing, as from our Discord discussion it feels like we're both leaning towards WebView2 being a better approach for custom tabs.

I'm definitely not against having both - feel free to reopen if this is something you want to pick up again.

**Pull request closed without merging**